File: algorithms/05_csharp/geometry/closest_pair.cs
using System;

namespace Algorithms.Geometry
{
    public static class ClosestPair
    {
        public static double Closest(double[][] points)
        {
            double best = double.MaxValue;
            for (int i = 0; i < points.Length; i++)
            {
                for (int j = i + 1; j < points.Length; j++)
                {
                    best = Math.Min(best, Math.Sqrt(
                        Math.Pow(points[i][0] - points[j][0], 2) + Math.Pow(points[i][1] - points[j][1], 2)));
                }
            }
            return best;
        }

        public static void Main()
        {
            var points = new[] { new[] { 2.0, 3.0 }, new[] { 12.0, 30.0 }, new[] { 40.0, 50.0 }, new[] { 5.0, 1.0 }, new[] { 12.0, 10.0 }, new[] { 3.0, 4.0 } };
            if (Math.Abs(Closest(points) - Math.Sqrt(2.0)) > 1e-9)
            {
                throw new InvalidOperationException("wrong closest distance");
            }
            Console.WriteLine($"closest={Closest(points)}");
        }
    }
}

File: algorithms/04_java/geometry/ClosestPair.java
public final class ClosestPair {
    private ClosestPair() {}

    public static double closest(double[][] pts) {
        double best = Double.MAX_VALUE;
        for (int i = 0; i < pts.length; i++) {
            for (int j = i + 1; j < pts.length; j++) {
                best = Math.min(best, Math.hypot(pts[i][0] - pts[j][0], pts[i][1] - pts[j][1]));
            }
        }
        return best;
    }

    public static void main(String[] args) {
        double[][] pts = {{2, 3}, {12, 30}, {40, 50}, {5, 1}, {12, 10}, {3, 4}};
        if (Math.abs(closest(pts) - Math.sqrt(2)) > 1e-9) {
            throw new AssertionError("wrong closest distance");
        }
        System.out.println("closest=" + closest(pts));
    }
}
